File: server/src/image_msg.rs
use super::message;
use message::{Message, MessageId, RecvMessage, SendMessage};
use std::any::Any;

#[derive(Debug)]
pub struct CaptureImageMsg {
    pub id: u8,
    pub camera_id: u8,
    data: Vec<u8>,
}

impl CaptureImageMsg {
    pub fn new() -> CaptureImageMsg {
        let id_value = MessageId::CaptureImage as u8;
        CaptureImageMsg {
            id: id_value,
            camera_id: 0,
            data: Vec::new(),
        }
    }
}

impl Message for CaptureImageMsg {
    fn id(&self) -> u8 {
        return self.id;
    }

    fn as_any(&self) -> &dyn Any {
        self
    }

    fn as_mut_any(&mut self) -> &mut dyn Any {
        self
    }
}

impl SendMessage for CaptureImageMsg {
    fn size(&self) -> u32 {
        return 1 + 2 + 2;
    }

    fn to_bytes(&mut self) -> Option<&[u8]> {
        self.data.push(self.camera_id);
        return Some(&self.data);
    }
}

#[derive(Debug)]
pub struct RecvImageMsg {
    pub id: u8,
    pub camera_id: u8,
    pub channels: u16,
    pub frame_width: u16,
    pub frame_height: u16,
    pub data: Vec<u8>,
}

impl RecvImageMsg {
    pub fn new() -> RecvImageMsg {
        let id_value = MessageId::RecvImage as u8;
        RecvImageMsg {
            id: id_value,
            camera_id: 0,
            channels: 0,
            frame_width: 0,
            frame_height: 0,
            data: Vec::new(),
        }
    }
}

impl Message for RecvImageMsg {
    fn id(&self) -> u8 {
        return self.id;
    }

    fn as_any(&self) -> &dyn Any {
        self
    }

    fn as_mut_any(&mut self) -> &mut dyn Any {
        self
    }
}

impl RecvMessage for RecvImageMsg {
    fn from_bytes(&mut self, _buf: &[u8]) {
        {
            self.camera_id = _buf[0];
        }
        {
            let tmp = slice_as_array!(&_buf[1..3], [u8; 2]).expect("RecvImageMsg wrong data");
            self.channels = u16::from_be_bytes(*tmp);
        }
        {
            let tmp = slice_as_array!(&_buf[3..5], [u8; 2]).expect("RecvImageMsg wrong data");
            self.frame_width = u16::from_be_bytes(*tmp);
        }
        {
            let tmp = slice_as_array!(&_buf[5..7], [u8; 2]).expect("RecvImageMsg wrong data");
            self.frame_height = u16::from_be_bytes(*tmp);
        }
        self.data.extend_from_slice(&_buf[7..]);
    }
}

File: client/image_msg.py
from message import MessageId, RecvMessage, SendMessage


class CaptureImageMsg(RecvMessage):
    def __init__(self):
        super(CaptureImageMsg, self).__init__(MessageId.CAPTURE_IMAGE)
        self.camera_id = 0
        self.img_width = 0
        self.img_height = 0

    def from_bytes(self, data):
        self.camera_id = int.from_bytes(data[0:1], byteorder='big')
        self.img_width = int.from_bytes(data[1:3], byteorder='big')
        self.img_height = int.from_bytes(data[3:5], byteorder='big')


class SendImageMsg(SendMessage):
    def __init__(self):
        super(SendImageMsg, self).__init__(MessageId.SEND_IMAGE)
        self.camera_id = 0
        self.img_width = 0
        self.img_height = 0
        self.img_channels = 0
        self.encoded = 0

    def set_img(self, camera_id, data, channels, width, height, encoded):
        size = 1 + 1 + 2 + 2 + 2 + (width * height * channels)
        self.size_ = 1 + size
        self.camera_id = camera_id
        self.img_width = width
        self.img_height = height
        self.img_channels = channels
        self.encoded = encoded
        self.add_bytes(self.camera_id.to_bytes(1, byteorder="big"))
        self.add_bytes(self.encoded.to_bytes(1, byteorder="big"))
        self.add_bytes(self.img_channels.to_bytes(2, byteorder="big"))
        self.add_bytes(self.img_width.to_bytes(2, byteorder="big"))
        self.add_bytes(self.img_height.to_bytes(2, byteorder="big"))
        self.add_bytes(data)
